wow_message_parser/src/rust_printer/structs/print_common_impls/print_read.rs

```diff
@@ -220,28 +203,50 @@ fn print_array_ty(
     array: &Array,
     d: &StructMemberDefinition,
     prefix: &str,
+    reader: &str,
     postfix: &str,
 ) {
     match array.ty() {
         ArrayType::Integer(integer_type) => {
             s.wln(format!(
-                "{name}.push({module}::{prefix}read_{int_type}_{endian}(r){postfix}?);",
+                "{name}.push({module}::{prefix}read_{int_type}_{endian}({reader}){postfix}?);",
                 name = d.name(),
                 module = UTILITY_PATH,
                 int_type = integer_type.rust_str(),
                 endian = integer_type.rust_endian_str(),
                 prefix = prefix,
+                reader = reader,
                 postfix = postfix,
             ));
+
+            if array.size() == ArraySize::Endless {
+                s.wln("current_size += 1;")
+            }
         }
         ArrayType::Struct(c) => {
             s.wln(format!(
-                "{name}.push({type_name}::{prefix}read(r){postfix}?);",
-                name = d.name(),
+                "let o = {type_name}::{prefix}read({reader}){postfix}?;",
                 type_name = c.name(),
                 prefix = prefix,
+                reader = reader,
                 postfix = postfix,
             ));
+
+            if array.size() == ArraySize::Endless {
+                if c.is_constant_sized() {
+                    s.wln(format!(
+                        "current_size += std::mem::size_of::<{type_name}>();",
+                        type_name = c.name()
+                    ));
+                } else {
+                    s.wln("current_size += o.size();");
```

Comment:
This is a problem when decompressing endless arrays.

Because we loop until current_size > body_size, we need to increment `current_size` by the amount of `compressed` bytes we've read, not decompressed bytes. Unfortunately, flate2 does some internal buffering so `total_in` isn't accurate, and just returns the full byte count on the first loop, even though we've only read a few bytes.

One alternative would be to change the loop in endless array reads so that we just loop until the buffer is fully emptied, by checking if the reader is EOF? Not sure how clean this would be

Reply:
I think the best alternative is to read the bytes as and endless u8 array into a buffer and then read the structs from that buffer? That way we could reliably use EOF delimiter for the correct amount of bytes.

Reply:
I think I've found a nice workaround:
```rust
        while decoder.total_out() < (decompressed_addon_info_size as u64) {
```
Like we discussed, every compressed tag points at the field containing the decompressed size, and `total_out` is accurate to the number of bytes we've read out of the ZLibDecoder, so this works great.

Last thing for this PR is to figure out how to fix the test assertion that checks the size of the struct vs the raw data, which fails because it's comparing the decompressed data to the compressed data.

Reply:
I'm not entirely sure, but I don't think that the decompressed length is 100% accurate. We can try it out though and see if it works.

Reply:
It *seems* to work nicely, but I'm hitting another problem in the test with a mismatched header, so I'll get back to you on that. Will fall back to your approach if we can't rely on `decompressed_addon_info_size`.